src/toil/test/src/fileStoreTest.py

```diff
@@ -250,6 +252,40 @@ def _testFileStoreOperations(job, nonLocalDir, numIters=100):
                             logger.info('No longer have file: %s', fsID)
                 i += 1
 
+        def testWriteReadGlobalFile(self):
+            workdir = self._createTempDir(purpose='nonLocalDir')
+            for i in True,False:
+                A = Job.wrapJobFn(self._testWriteReadGlobalFilePermissions,
+                                isLocalFile=False, executable=i, nonLocalDir=workdir)
+                Job.Runner.startToil(A, self.options)
+
+        @staticmethod
+        def _testWriteReadGlobalFilePermissions(job, isLocalFile, executable, nonLocalDir=None):
+            """
+            Ensures that uploaded files preserve their file permissions when they
+            are downloaded again. This function checks that a written executable file
+            maintains its executability after being read.
+            """
+            if isLocalFile:
+                workDir = job.fileStore.getLocalTempDir()
+            else:
+                assert nonLocalDir is not None
+                workDir = nonLocalDir
+            srcFile = '%s/%s%s' % (workDir, 'in', str(uuid4()))
+            with open(srcFile, 'w') as f:
+                f.write('Hello')
+
+            if executable:
+                os.chmod(srcFile, os.stat(srcFile).st_mode | stat.S_IXUSR)
+
+            initialPermissions = os.stat(srcFile).st_mode
+            fileID = job.fileStore.writeGlobalFile(srcFile)
+            dstFile = job.fileStore.getLocalTempFile() + str(uuid4())
+            dstFile = job.fileStore.readGlobalFile(fileID, userPath=dstFile, mutable=True)
+            currentPermissions = os.stat(dstFile).st_mode
+
+            assert initialPermissions == currentPermissions
```

Comment:
You might not want to compare the whole mode, just the `stat.S_IXUSR` bit (`assert initialPermissions & stat.S_IXUSR == currentPermissions & stat.S_IXUSR`)